django/db/models/indexes.py

```diff
@@ -0,0 +1,118 @@
+from __future__ import unicode_literals
+
+import hashlib
+
+from django.core.exceptions import ValidationError
+from django.utils.encoding import force_bytes
+
+__all__ = ['Index']
+
+# The max length of the names of the indexes (restricted to 30 due to Oracle)
+MAX_NAME_LENGTH = 30
+
+
+class Index(object):
+
+    index_type = 'idx'
+
+    def __init__(self, *fields, **kwargs):
+        if not fields:
+            raise ValueError("Minimum one field is required to define an index.")
+        self.fields = fields
+        self._name = kwargs.get('name', '')
+        if len(self._name) > MAX_NAME_LENGTH:
+            raise ValidationError(
+                "Index names cannot be longer than %s chars." % MAX_NAME_LENGTH
+            )
+        if 'model' in kwargs:
+            self.model = kwargs['model']
```

Comment:
Any problem with allowing `self.model = None`. I think conditional attributes which require `hasattr` isn't the best design.